### ISSUE TITLE
Internal Range Overflow/Loop

### ISSUE THREAD

**Issue by 3FLLC** [nielsAD/lape#50]
var
   A:Byte;

for A:=1 to 255 do write(a," ");

output:
0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 ... 251 252 253 254 255 0

**Comment by nielsAD** [comment]
Yes, this a bug. Not sure what would be the best way to fix this. 

For completeness sake; PascalScript also has this issue :stuck_out_tongue_winking_eye:

Thanks for reporting!

**Comment by riwu** [comment]
Perhaps check if (i = targetValue) at the end of the loop before the increment?
And check if (i => targetValue) before executing the for loop.
This way the variable won't ever be incremented beyond the targetValue.
I guess that's how FPC does it?

**Comment by nielsAD** [comment]
FPC does it like this:
```
1: if (counter > limit) then goto 6
2: dec counter
3: inc counter
4: [LOOP BODY]
5: if (counter < limit) then goto 3
```
So I guess it can be solved by emitting two comparisons.

**Comment by 3FLLC** [comment]
Another example, this one just bit me BAD!

    Raises:  Runtime error: "Index out of range (index:4294967295, low:0, high:382)"

    procedure generatestatistics;
    var
        loop:longword;

    begin
        for loop:=high(filenames)-1 downto 0 do begin

-- it loops around :-( ... so I had to record using while loops. Can you point me where to put a bandaid for the iteration? (downto and to) and (step).

Thanks!
Ozz

PS. Happy Holidays!

**Comment by riwu** [comment]
Declare your loop control variable as integer (signed) instead of longword (unsigned).

**Comment by 3FLLC** [comment]
Does no good if my low end goes below the range - only solution is what Niels mentioned. Thanks, but no Band-aids are acceptable. ;-)

**Comment by riwu** [comment]
Your low end goes below -2^32? How many files do you have? You can use Int64 for bigger range.
You asked for a band-aid, but no band-aids are acceptable?

**Comment by 3FLLC** [comment]
I meant for the “band-aid” to the actual LA-PE engine, not for the script – I recoded to use while loops. (sorry for lack of clarity on the band-aid type). Just frustrated because this was a production project then POW. :-(

**Comment by nielsAD** [comment]
The `downto` example is something different. As @riwu said, this is because you are assigning a negative number to an unsigned variable. [Range checking](http://www.freepascal.org/docs-html/3.0.0/prog/progsu65.html#x72-710001.2.65) before assigning integer types could catch this, but that's unrelated to this for-loops issue.
